cli/src/commands/bookmark/list.rs

```diff
@@ -277,3 +307,492 @@ struct RefListItem {
+            SortKey::CommitterDate => bookmark_items
+                .sort_by_key(|item| to_commit(item).map(|commit| commit.committer.timestamp)),
+            SortKey::CommitterDateDesc => bookmark_items.sort_by_key(|item| {
+                cmp::Reverse(to_commit(item).map(|commit| commit.committer.timestamp))
+            }),
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use std::collections::HashMap;
+
+    use backend::ChangeId;
+    use backend::MergedTreeId;
+    use backend::TreeId;
+    use jj_lib::backend::MillisSinceEpoch;
+    use jj_lib::backend::Signature;
+    use jj_lib::backend::Timestamp;
+    use jj_lib::op_store::RefTarget;
+    use jj_lib::op_store::RemoteRef;
+    use jj_lib::op_store::RemoteRefState;
+
+    use super::*;
+
+    fn make_backend_commit(author: Signature, committer: Signature) -> Arc<backend::Commit> {
+        Arc::new(backend::Commit {
+            parents: vec![],
+            predecessors: vec![],
+            root_tree: MergedTreeId::Legacy(TreeId::new(vec![])),
+            change_id: ChangeId::new(vec![]),
+            description: String::new(),
+            author,
+            committer,
+            secure_sig: None,
+        })
+    }
+
+    fn make_default_signature() -> Signature {
+        Signature {
+            name: "Test User".to_owned(),
+            email: "test.user@example.com".to_owned(),
+            timestamp: Timestamp {
+                timestamp: MillisSinceEpoch(0),
+                tz_offset: 0,
+            },
+        }
+    }
+
+    fn commit_id_generator() -> impl FnMut() -> CommitId {
+        let mut iter = (1_u128..).map(|n| CommitId::new(n.to_le_bytes().into()));
+        move || iter.next().unwrap()
+    }
+
+    fn commit_ts_generator() -> impl FnMut() -> Timestamp {
+        // iter starts as 1, 1, 2, ... for test purposes
+        let mut iter = Some(1_i64).into_iter().chain(1_i64..).map(|ms| Timestamp {
+            timestamp: MillisSinceEpoch(ms),
+            tz_offset: 0,
+        });
+        move || iter.next().unwrap()
+    }
+
+    // Helper function to prepare test data, sort and prepare snapshot with relevant
+    // information.
+    fn prepare_data_sort_and_snapshot(sort_keys: &[SortKey]) -> String {
+        let mut new_commit_id = commit_id_generator();
+        let mut new_timestamp = commit_ts_generator();
+        let names = ["bob", "alice", "eve", "bob"];
+        let emails = ["bob@g.com", "alice@g.com", "eve@g.com", "bob@g.com"];
+        let bookmark_names = ["feature", "bug-fix", "chore", "bug-fix"];
+        let remote_names = [None, Some("upstream"), None, Some("origin")];
+        let mut bookmark_items: Vec<RefListItem> = Vec::new();
+        let mut commits: HashMap<CommitId, Arc<backend::Commit>> = HashMap::new();
+        for (((&name, &email), bookmark_name), remote_name) in names
+            .iter()
+            .zip(emails.iter())
+            .zip(bookmark_names.iter())
+            .zip(remote_names.iter())
+        {
```

Comment:
nit: `itertools::izip!(&names, &emails, &bookmark_names, &remote_names)` can be used